app/controllers/internal_api/v1/pillars_controller.rb

```diff
@@ -106,16 +106,30 @@ def ec2_cloud_contents
   end
 
   def azure_cloud_contents
+    tenant = Pillar.value(pillar: :azure_tenant_id)
+    client = Pillar.value(pillar: :azure_client_id)
+    secret = Pillar.value(pillar: :azure_secret)
+
+    if tenant && client && secret # rubocop:disable Style/ConditionalAssignment
+      # Using service principal credentials for authentication
+      azure_provider = {
+        subscription_id: Pillar.value(pillar: :azure_subscription_id),
+        tenant:          tenant,
+        client_id:       client,
+        secret:          secret
+      }
+    else
+      # Using MSI authorization
+      azure_provider = {
+        subscription_id: Pillar.value(pillar: :azure_subscription_id)
+      }
+    end
```

Comment:
Even though this working I wonder if it would be better to extract 

```
azure_provider = {
    subscription_id: Pillar.value(pillar: :azure_subscription_id)
}
```

and use only one `if` to merge the `tenant`, `client_id` and `secret` if those exist.